### PR TITLE
Update Channels integration

### PR DESCRIPTION
**Pull request by cdroege**
## Description

This updates the Django Channels integration to use the base classes that are used by all other integrations. For this, I had to add a new custom `ChannelsRequest` class to hold the needed information (i.e. body and consumer) to handle a HTTP request.

This also adds two features:

1. Allow setting response headers
2. Add `multipart/form-data` support so that file uploads work

## Types of Changes

- [ ] Core
- [ ] Bugfix
- [x] New feature
- [x] Enhancement/optimization
- [x] Documentation

## Issues Fixed or Closed by This PR

* #2764
* #2694

## Checklist

- [x] My code follows the code style of this project.
- [x] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.
- [x] I have read the CONTRIBUTING document.
- [x] I have added tests to cover my changes.
- [x] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

### CODE REVIEW

**Comment by cdroege** [comment]
All tests are green now.

@patrick91 I have one question regarding the creation of a context in `get_context`. Should this still use `StrawberryChannelsContext` or should it use a `dict` like the other integrations? In your commit (acd48de3f8ed84b9e8de74df909dada194d1da45) you changed it to a dict, but I am unsure, if this should be changed as this breaks backwards compatibility when accessing a context (see the test change in b58a3638206e62e8a6f11ec54d87a0acf00a9cf3).

**Comment by patrick91** [comment]
> All tests are green now.
> 
> @patrick91 I have one question regarding the creation of a context in `get_context`. Should this still use `StrawberryChannelsContext` or should it use a `dict` like the other integrations? In your commit ([acd48de](https://github.com/strawberry-graphql/strawberry/commit/acd48de3f8ed84b9e8de74df909dada194d1da45)) you changed it to a dict, but I am unsure, if this should be changed as this breaks backwards compatibility when accessing a context (see the test change in [b58a363](https://github.com/strawberry-graphql/strawberry/commit/b58a3638206e62e8a6f11ec54d87a0acf00a9cf3)).

let's keep it as it is for now :) we have this issue: #2687 to move request and response to info so we don't have to rely on the context 😊

**Comment by cdroege** [comment]
I tested the changes with my code base and it works :-)

I just need to add some more documentation (a RELEASE.md file, documentation about the breaking changes regarding `get_context` and I want to add an example for overwriting `get_context` so that a user coming from an older release knows how to migrate to a custom context dataclass, if a `dict` is not enough).

But I have two questions left, @patrick91:

1: It is probably fine, if I add a `# pragma: no cover` to the two lines not covered by tests? https://app.codecov.io/gh/strawberry-graphql/strawberry/pull/2775/blob/strawberry/channels/handlers/http_handler.py
2. The release type is `minor`?

**Comment by botberry** [comment]
Thanks for adding the `RELEASE.md` file!



Here's a preview of the changelog:

---

This release updates the Django Channels integration so that it uses the same base
classes used by all other integrations.

**New features:**

The Django Channels integration supports two new features:

* Setting headers in a response
* File uploads via `multipart/form-data` POST requests

**Breaking changes:**

This release contains a breaking change for the Channels integration. The context
object is now a `dict` and it contains different keys depending on the connection
protocol:

1. HTTP: `request` and `response`. The `request` object contains the full
   request (including the body). Previously, `request` was the `GraphQLHTTPConsumer`
   instance of the current connection. The consumer is now available via
   `request.consumer`.
2. WebSockets: `request`, `ws` and `response`. `request` and `ws` are the same
   `GraphQLWSConsumer` instance of the current connection.

If you want to use a dataclass for the context object (like in previous releases),
you can still use them by overriding the `get_context` methods. See the Channels
integration documentation for an example.


---
Here's the preview release card for twitter:




Here's the tweet text: 
```
🆕 Release (next) is out! Thanks to Christian Dröge for the PR 👏

Get it here 👉 https://github.com/strawberry-graphql/strawberry/releases/tag/(next)
```

**Comment by patrick91** [comment]
@cdroege I'll start reviewing this now, thanks so much for this!

File: docs/breaking-changes.md
Comment (patrick91) [inline docs/breaking-changes.md:7]: 👏 I'll update this before merging it 😊

File: docs/breaking-changes/0.179.0.md
Comment (patrick91) [inline docs/breaking-changes/0.179.0.md:21]: this is well written!

File: tests/http/conftest.py
Comment (patrick91) [inline tests/http/conftest.py:39]: thanks for the comment!

**Review by patrick91 — APPROVED** [review]
This is amazing! Thank you so much! <3

**Comment by patrick91** [comment]
not sure what pre-commit.ci is complaining about, I'll merge it anyway and see if I get the same error on main

Thank you so much for this!

**Comment by botberry** [comment]
Thanks for contributing to Strawberry! 🎉 You've been invited to join
the Strawberry GraphQL organisation 😊

You can also request a free sticker by filling this form: https://forms.gle/dmnfQUPoY5gZbVT67

And don't forget to join our discord server: https://strawberry.rocks/discord 🔥

**Pull request merged**
